File: Aircraft/Aircraft.py
import csv
import inspect
import json
from pathlib import Path

from aircraft_cpp import CppAircraft
from Armaments.Armament import Armament, load_armament_type
from Countermeasures.Countermeasure import Countermeasure, load_countermeasure_type

DATA_ROOT = Path(__file__).resolve().parent


class Aircraft:
    def __init__(self, name, manufacturer):

        self.name = name
        self.manufacturer = manufacturer
        self.path = Path(inspect.getfile(type(self))).resolve().parent
        self.data_path = self.path / "data"

        self.radar = self._load_json("radar.json")
        self.ir_sensor = self._load_json("ir_sensor.json")
        self.radio = self._load_json("radio.json")

        self.rcs_signature = self._load_table_csv(self.data_path / "rcs_signature.csv")
        self.ir_signature_tables = self._load_table_dir("ir_signature")

        self.fuel_tables = self._load_table_dir("fuel")
        self.temperature_tables = self._load_table_dir("temperature")
        self.temperatures_c = {}  # mirrors the C++ backend's state, for external inspection

        self.armaments: dict[Armament, int] = self._load_loadout("armaments", load_armament_type)
        self.countermeasures: dict[Countermeasure, int] = self._load_loadout("countermeasures", load_countermeasure_type)

        # Temperature dynamics and sensing math (interpolation, sector containment,
        # aspect matching, Stefan-Boltzmann scaling, range falloff) all live in C++ now
        # for speed -- see Aircraft/capi.h and aircraft_cpp.py. Python still owns all the
        # file I/O above; this just mirrors the parsed data into the C++ backend.
        self._cpp = self._create_cpp_backend()
        self._sync_cpp_backend()

    def _create_cpp_backend(self):
        """Every concrete aircraft type must override this to return its matching
        aircraft_cpp.CppAircraft (e.g. CppAircraft.f16()) -- see F16.py/Aim120.py."""
        raise NotImplementedError(f"{type(self).__name__} has no C++ backend wired up (see aircraft_cpp.py)")

    def _sync_cpp_backend(self):
        fov = self.radar["field_of_regard_deg"]
        self._cpp.set_radar_config(self.radar["max_range_km"], fov["azimuth"], fov["elevation"])

        for sensor_name, cfg in self.ir_sensor.items():
            ir_fov = cfg["field_of_view_deg"]
            self._cpp.set_ir_sensor_config(sensor_name, cfg["max_range_km"], ir_fov["azimuth"], ir_fov["elevation"])

        for table_name, rows in self.temperature_tables.items():
            x_key = next(k for k in rows[0] if k != "temp_c")
            self._cpp.set_temperature_table(table_name, rows, x_key)

        self._cpp.set_rcs_signature(self.rcs_signature)
        for component_name, rows in self.ir_signature_tables.items():
            self._cpp.set_ir_signature_component(component_name, rows)

    def _load_json(self, filename):
        with open(self.data_path / filename) as f:
            return json.load(f)

    def _load_table_dir(self, dirname):
        """One entry per CSV in data/<dirname>/, keyed by filename (e.g. data/fuel/fuel_to_thrust.csv -> "fuel_to_thrust")."""
        table_dir = self.data_path / dirname
        if not table_dir.is_dir():
            return {}
        return {
            csv_file.stem: self._load_table_csv(csv_file)
            for csv_file in sorted(table_dir.glob("*.csv"))
        }

    def _load_table_csv(self, path):
        with open(path, newline="") as f:
            reader = csv.DictReader(f)
            return [{k: _try_float(v) for k, v in row.items()} for row in reader]

    def _load_loadout(self, key, type_loader):
        """Reads data/loadout.json's "armaments"/"countermeasures" list (type name + count) into
        {instance: count}, resolving each type name to its class via type_loader."""
        loadout_file = self.data_path / "loadout.json"
        if not loadout_file.is_file():
            return {}
        with open(loadout_file) as f:
            loadout = json.load(f)
        return {
            type_loader(entry["type"])(): entry["count"]
            for entry in loadout.get(key, [])
        }

    def update_temperature(self, table_name, x_value, dt, time_constant_s=10.0):
        """Drifts this component's temperature toward data/temperature/<table_name>.csv's
        interpolated target at x_value (e.g. update_temperature("engine_core", 80, dt)),
        via the C++ backend's exact exponential heating/cooling curve. Returns (and mirrors
        into self.temperatures_c) the new current temperature.
        """
        current_c = self._cpp.update_temperature(table_name, x_value, dt, time_constant_s)
        self.temperatures_c[table_name] = current_c
        return current_c

    def radar_detect(self, t, sensor_position, sensor_orientation,
                      target, target_position, target_orientation):
        """Radar detection against `target` (another Aircraft/Missile). position is global
        XYZ in meters; orientation is a (w, x, y, z) quaternion, local -> world, same
        convention as the C++ physics code. See Aircraft::radarDetect in Aircraft.cpp for
        the full rationale (bounding-box-first, then nearest-aspect RCS lookup).
        """
        return self._cpp.radar_detect(t, sensor_position, sensor_orientation,
                                       target._cpp, target_position, target_orientation)

    def ir_detect(self, t, sensor_position, sensor_orientation,
                  target, target_position, target_orientation):
        """IR detection against `target`. See Aircraft::irDetect in Aircraft.cpp for the
        full rationale (bounding-box-first, then per-component nearest-aspect lookup summed
        and Stefan-Boltzmann-scaled by that component's current temperature).
        """
        return self._cpp.ir_detect(t, sensor_position, sensor_orientation,
                                    target._cpp, target_position, target_orientation)

    def perceive(self, signature, fields, sensor_strength=1.0, falloff_exponent=2.0):
        """Turns a raw signature (radar_detect's or ir_detect's return value, which
        includes "range_km") into what this aircraft's own sensor actually picks up:
        range dropoff combined with this sensor's own strength/sensitivity multiplier,
        via the C++ backend. falloff_exponent=4 for radar (round-trip radar equation),
        =2 (the default) for IR (one-way, inverse-square).
        """
        range_km = signature["range_km"]
        return {
            field: CppAircraft.perceive(signature[field], range_km, sensor_strength, falloff_exponent)
            for field in fields
        }


def _try_float(value):
    try:
        return float(value)
    except ValueError:
        return value

File: Aircraft/Aircraft.cpp
#include "Aircraft.h"

#include <cmath>

Aircraft::Aircraft(std::string name, std::string manufacturer)
    : name_(std::move(name)), manufacturer_(std::move(manufacturer))
{
}

void Aircraft::applyGravity(glm::vec3& position, glm::quat& /*orientation*/,
                             glm::vec3& momentum, const glm::vec3& /*thrust*/, float dt)
{
    momentum.y -= massKg * kGravity * dt;
    position += (momentum / massKg) * dt;
}

void Aircraft::updateMass()
{
    massKg = emptyMassKg + fuelMassKg + armamentsMassKg + countermeasuresMassKg;
}

void Aircraft::updatePhysics(glm::vec3& position, glm::quat& orientation,
                              glm::vec3& momentum, const glm::vec3& thrust, float dt)
{
    applyGravity(position, orientation, momentum, thrust, dt);
}

float Aircraft::updateTemperature(const std::string& tableName, float xValue, float dt, float timeConstantS)
{
    float targetC = interp1d(temperatureTables.at(tableName), xValue);
    float currentC = kAmbientTempC;
    auto it = temperaturesC.find(tableName);
    if (it != temperaturesC.end()) {
        currentC = it->second;
    }
    currentC += (targetC - currentC) * (1.0f - std::exp(-dt / timeConstantS));
    temperaturesC[tableName] = currentC;
    return currentC;
}

SectorVolume Aircraft::staticSector(const std::string& sectorName, const FovConfig& fov)
{
    SectorVolume v;
    v.name = sectorName;
    v.centerAzimuthDeg = 0.0f;
    v.azimuthWidthDeg = fov.azimuthDeg;
    v.centerElevationDeg = 0.0f;
    v.elevationWidthDeg = fov.elevationDeg;
    v.rangeKm = fov.maxRangeKm;
    return v;
}

std::vector<SectorVolume> Aircraft::radarCoverage(float /*t*/) const
{
    return {staticSector("primary", radarConfig)};
}

std::vector<SectorVolume> Aircraft::irCoverage(float /*t*/) const
{
    std::vector<SectorVolume> result;
    for (const auto& [sensorName, fov] : irSensorConfigs) {
        result.push_back(staticSector(sensorName, fov));
    }
    return result;
}

Aircraft::DetectionResult Aircraft::radarDetect(float t, const glm::vec3& sensorPosition,
                                                 const glm::quat& sensorOrientation, const Aircraft& target,
                                                 const glm::vec3& targetPosition,
                                                 const glm::quat& targetOrientation) const
{
    glm::vec3 delta = targetPosition - sensorPosition;
    float rangeKm = glm::length(delta) / 1000.0f;

    glm::vec3 sensorLocalDelta = glm::conjugate(sensorOrientation) * delta;
    auto [bearingDeg, elevationDeg] = bearingElevationDeg(sensorLocalDelta);

    bool inCoverage = false;
    for (const auto& volume : radarCoverage(t)) {
        if (volume.contains(bearingDeg, elevationDeg, rangeKm)) {
            inCoverage = true;
            break;
        }
    }
    if (!inCoverage) {
        return DetectionResult{};
    }

    glm::vec3 targetLocalDelta = glm::conjugate(targetOrientation) * delta;
    glm::vec3 aspectDir = glm::normalize(targetLocalDelta);

    const SignatureRow& row = nearestSignatureRow(target.rcsSignature, aspectDir);

    DetectionResult result;
    result.hit = true;
    result.values = row.values;
    result.values["range_km"] = rangeKm;
    result.values["aspect_x"] = row.direction.x;
    result.values["aspect_y"] = row.direction.y;
    result.values["aspect_z"] = row.direction.z;
    return result;
}

Aircraft::DetectionResult Aircraft::irDetect(float t, const glm::vec3& sensorPosition,
                                              const glm::quat& sensorOrientation, const Aircraft& target,
                                              const glm::vec3& targetPosition,
                                              const glm::quat& targetOrientation) const
{
    glm::vec3 delta = targetPosition - sensorPosition;
    float rangeKm = glm::length(delta) / 1000.0f;

    glm::vec3 sensorLocalDelta = glm::conjugate(sensorOrientation) * delta;
    auto [bearingDeg, elevationDeg] = bearingElevationDeg(sensorLocalDelta);

    bool inCoverage = false;
    for (const auto& volume : irCoverage(t)) {
        if (volume.contains(bearingDeg, elevationDeg, rangeKm)) {
            inCoverage = true;
            break;
        }
    }
    if (!inCoverage) {
        return DetectionResult{};
    }

    glm::vec3 targetLocalDelta = glm::conjugate(targetOrientation) * delta;
    glm::vec3 aspectDir = glm::normalize(targetLocalDelta);

    float totalMwir = 0.0f;
    float totalLwir = 0.0f;
    for (const auto& [componentName, table] : target.irSignatureTables) {
        const SignatureRow& row = nearestSignatureRow(table, aspectDir);

        const auto& tempTable = target.temperatureTables.at(componentName);
        float referenceTempC = tempTable.front().second;
        for (const auto& point : tempTable) {
            referenceTempC = std::max(referenceTempC, point.second);
        }

        float currentTempC = kAmbientTempC;
        auto it = target.temperaturesC.find(componentName);
        if (it != target.temperaturesC.end()) {
            currentTempC = it->second;
        }

        float scale = irTemperatureScale(currentTempC, referenceTempC);
        totalMwir += row.values.at("ir_mwir_w_per_sr") * scale;
        totalLwir += row.values.at("ir_lwir_w_per_sr") * scale;
    }

    DetectionResult result;
    result.hit = true;
    result.values["ir_mwir_w_per_sr"] = totalMwir;
    result.values["ir_lwir_w_per_sr"] = totalLwir;
    result.values["range_km"] = rangeKm;
    return result;
}

std::unordered_map<std::string, float> Aircraft::perceive(const std::unordered_map<std::string, float>& signature,
                                                            const std::vector<std::string>& fields,
                                                            float sensorStrength, float falloffExponent) const
{
    float rangeKm = signature.at("range_km");
    float divisor = rangeKm <= 0.0f ? 1.0f : std::pow(rangeKm, falloffExponent);
    std::unordered_map<std::string, float> result;
    for (const auto& field : fields) {
        result[field] = signature.at(field) * sensorStrength / divisor;
    }
    return result;
}
